meta/client.go

```diff
@@ -0,0 +1,224 @@
+package meta
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/XiaoMi/pegasus-go-client/idl/base"
+	"github.com/XiaoMi/pegasus-go-client/session"
+	"github.com/bluele/gcache"
+	"github.com/go-zookeeper/zk"
+	"github.com/sirupsen/logrus"
+)
+
+// TODO(jiashuo) store config file
+var zkAddrs = []string{""}
+var zkTimeOut = 1000000000 // unit ns, equal 1s
+var zkRoot = "/pegasus-cluster"
+var zkWatcherCount = 1024
+
+var globalClusterManager *ClusterManager
+
+type ClusterManager struct {
+	Mut    sync.RWMutex
+	ZkConn *zk.Conn
+	// table->TableInfoWatcher
+	Tables gcache.Cache
+	// metaAddrs->metaManager
+	Metas map[string]*session.MetaManager
+}
+
+// zkContext cancels the goroutine that's watching the zkNode, when the watcher
+// is evicted by LRU.
+type zkContext struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type TableInfoWatcher struct {
+	tableName   string
+	clusterName string
+	metaAddrs   string
+	event       <-chan zk.Event
+	ctx         zkContext
+}
+
+// TODO(jishuo1) change log module
+func initClusterManager() {
+	zkConn, _, err := zk.Connect(zkAddrs, time.Duration(zkTimeOut))
+	if err != nil {
+		logrus.Panicf("failed to connect to zookeeper \"%s\": %s", zkAddrs, err)
+	}
+
+	tables := gcache.New(zkWatcherCount).LRU().EvictedFunc(func(key interface{}, value interface{}) {
+		value.(*TableInfoWatcher).ctx.cancel()
+		logrus.Warnf("[%s] evicted from table cache (capacity: %d)", key.(string), zkWatcherCount)
+	}).Build() // TODO(jiashuo1) consider set expire time
+	globalClusterManager = &ClusterManager{
+		ZkConn: zkConn,
+		Tables: tables,
+		Metas:  make(map[string]*session.MetaManager),
+	}
+
+	logrus.Infof("init cluster manager: zk=%s, zkTimeOut=%d(ms), zkRoot=%s, zkWatcherCount=%d",
+		zkAddrs, zkTimeOut/(1000*1000), zkRoot, zkWatcherCount)
+}
+
+func (m *ClusterManager) getMeta(table string) (*session.MetaManager, error) {
+	var meta *session.MetaManager
+
+	tableInfo, err := m.Tables.Get(table)
+	if err == nil {
+		meta = m.Metas[tableInfo.(*TableInfoWatcher).metaAddrs]
+		if meta != nil {
+			return meta, nil
+		}
+	}
+
+	logrus.Infof("[%s] can't get cluster info from local cache, try fetch from zk.", table)
+	m.Mut.Lock()
+	defer m.Mut.Unlock()
+	tableInfo, err = m.Tables.Get(table)
+	if err != nil {
+		tableInfo, err = m.newTableInfo(table)
+		if err != nil {
+			logrus.Errorf("[%s] get cluster info failed: %s", table, err)
+			return nil, err
+		}
+		err = m.Tables.Set(table, tableInfo)
+		if err != nil {
+			logrus.Errorf("[%s] local cache cluster info is updated failed: %s", table, err)
+			return nil, base.ERR_INVALID_DATA
+		}
+	}
+
+	metaAddrs := tableInfo.(*TableInfoWatcher).metaAddrs
```

Comment:
```suggestion
        tableInfoW := tableInfo.(*TableInfoWatcher)
	metaAddrs := tableInfoW.metaAddrs
```

下面的 `tableInfo.(*TableInfoWatcher)` 也都改掉